File: orchestrator/src/storage/redis/scripts/add_instance.lua
-- SPDX-FileCopyrightText: OpenTalk GmbH <mail@opentalk.eu>
--
-- SPDX-License-Identifier: EUPL-1.2

-- Returns a pair of `{ code, conflicting_resources }`, where the code is one of the values
-- documented on `ScriptResult` in `add_instance.rs`. `conflicting_resources` is only populated for
-- the resource conflict code and empty otherwise.

local instance_key = KEYS[1]
local instance_metrics_key = KEYS[2]
local instance_resource_key = KEYS[3]
local service_kind_key = KEYS[4]
local orchestrator_services_key = KEYS[5]

local service_id = ARGV[1]
local url = ARGV[2]
local kind = ARGV[3]
local key_ids = ARGV[4]
local load = ARGV[5]
local accepting_jobs = ARGV[6]
local managed_resources = ARGV[7]
local resource_keys = ARGV[8]

-- Check if a service instance with the same service id already exists

if redis.call("EXISTS", instance_key) == 1 then
    return { 1, {} } -- service instance already exists
end

local resources = cjson.decode(managed_resources)
local resource_lookup_keys = cjson.decode(resource_keys)

-- A list of resources already managed by another service instance
local conflicts = {}

for i, resource in ipairs(resources) do
    if redis.call("EXISTS", resource_lookup_keys[i]) == 1 then
        conflicts[#conflicts + 1] = resource
    end
end

if #conflicts > 0 then
    return { 2, conflicts } -- resources are owned by other instances
end

-- Claim the resources
for i, resource in ipairs(resources) do
    redis.call("SET", resource_lookup_keys[i], service_id)
    redis.call("SADD", instance_resource_key, resource)
end

-- Set service data
redis.call("HSET", instance_key,
            "url", url,
            "kind", kind,
            "api_key_ids", key_ids)

-- Set metrics
redis.call("HSET", instance_metrics_key,
            "load", load,
            "accepting_jobs", accepting_jobs)

-- Add to service set
redis.call("SADD", service_kind_key, service_id)

-- Track which orchestrator owns this service (used for cleanup on crash)
redis.call("SADD", orchestrator_services_key, service_id)

return { 0, {} }

File: orchestrator/src/storage/redis/scripts/add_instance.rs
// SPDX-FileCopyrightText: OpenTalk GmbH <mail@opentalk.eu>
//
// SPDX-License-Identifier: EUPL-1.2

use anyhow::Context;
use opentalk_orchestrator_shared::{ServiceKind, error::RegistrationError};
use opentalk_service_auth::ApiKeyId;
use redis::{FromRedisValue, RedisError, ToRedisArgs};
use url::Url;
use uuid::Uuid;

use crate::storage::{
    ServiceResource,
    redis::keys::{
        OrchestratorServicesKey, ResourceLookupKey, ServiceId, ServiceInstanceKey, ServiceKindKey,
        ServiceMetricsKey, ServiceResourceKey,
    },
};

#[derive(Debug)]
enum ScriptResult {
    Ok = 0,
    AddressAlreadyInUse = 1,
    ResourceAlreadyExists = 2,
}

impl FromRedisValue for ScriptResult {
    fn from_redis_value(v: redis::Value) -> Result<Self, redis::ParsingError> {
        let int_value = i32::from_redis_value(v)?;
        match int_value {
            0 => Ok(ScriptResult::Ok),
            1 => Ok(ScriptResult::AddressAlreadyInUse),
            2 => Ok(ScriptResult::ResourceAlreadyExists),
            other => Err(format!("unexpected add_instance script return value: {other}").into()),
        }
    }
}

#[derive(Debug, thiserror::Error)]
pub(crate) enum AddInstanceScriptError {
    #[error("{0}")]
    Registration(RegistrationError),
    #[error("{0}")]
    Redis(#[from] RedisError),
    #[error("{0}")]
    Internal(#[from] anyhow::Error),
}

impl From<ScriptResult> for Result<(), AddInstanceScriptError> {
    fn from(value: ScriptResult) -> Self {
        match value {
            ScriptResult::Ok => Ok(()),
            ScriptResult::AddressAlreadyInUse => Err(AddInstanceScriptError::Registration(
                RegistrationError::AddressAlreadyInUse,
            )),
            ScriptResult::ResourceAlreadyExists => Err(AddInstanceScriptError::Registration(
                RegistrationError::ResourceAlreadyExists,
            )),
        }
    }
}

const SCRIPT: &str = include_str!("add_instance.lua");

/// Add a new service instance to the orchestrator cluster.
///
/// The script will initialize all relevant keys and check for conflicts with existing instances or
/// resources.
#[allow(clippy::too_many_arguments)]
pub(crate) async fn add_instance(
    client: &redis::Client,
    url: Url,
    kind: ServiceKind,
    key_ids: Vec<ApiKeyId>,
    load: u8,
    accepting_jobs: bool,
    managed_resources: Vec<ServiceResource>,
    orchestrator_id: Uuid,
) -> Result<(), AddInstanceScriptError> {
    let script = redis::Script::new(SCRIPT);
    let service_id = ServiceId::from(&url);

    let key_ids = serde_json::to_string(
        &key_ids
            .iter()
            .map(|key_id| key_id.to_string())
            .collect::<Vec<String>>(),
    )
    .context("Failed to serialize key ids")?;

    let mut resource_values = Vec::with_capacity(managed_resources.len());
    let mut resource_keys = Vec::with_capacity(managed_resources.len());

    for resource in &managed_resources {
        resource_values.push(resource.to_string());

        let resource_lookup_key = ResourceLookupKey {
            resource: *resource,
        };

        resource_keys.push(
            String::from_utf8_lossy(
                resource_lookup_key
                    .to_redis_args()
                    .first()
                    .context("Failed to build resource lookup key")?,
            )
            .to_string(),
        );
    }

    debug_assert_eq!(resource_keys.len(), managed_resources.len());

    let resource_keys = serde_json::to_string(&resource_keys)
        .context("Failed to serialize resource lookup keys")?;
    let resource_values =
        serde_json::to_string(&resource_values).context("Failed to serialize resource values")?;
    let mut con = client.get_multiplexed_async_connection().await?;

    let result = script
        .key(ServiceInstanceKey {
            id: service_id.clone(),
        })
        .key(ServiceMetricsKey {
            id: service_id.clone(),
        })
        .key(ServiceResourceKey {
            id: service_id.clone(),
        })
        .key(ServiceKindKey { kind })
        .key(OrchestratorServicesKey {
            id: orchestrator_id.to_string(),
        })
        .arg(service_id.as_str())
        .arg(url.as_str())
        .arg(kind.to_string())
        .arg(key_ids)
        .arg(load)
        .arg(accepting_jobs)
        .arg(resource_values)
        .arg(resource_keys)
        .invoke_async::<ScriptResult>(&mut con)
        .await?;

    result.into()
}
